crates/sparrow-sources/src/in_memory.rs

```diff
@@ -0,0 +1,221 @@
+
+use arrow_array::RecordBatch;
+use arrow_schema::SchemaRef;
+use error_stack::{IntoReportCompat, ResultExt};
+use futures::{Stream, StreamExt, TryStreamExt};
+
+use sparrow_batch::Batch;
+use sparrow_core::ReadConfig;
+use sparrow_interfaces::Source;
+use sparrow_merge::old::homogeneous_merge;
+
+use crate::Error;
+
+/// Wraps the mutable in-memory batch container.
+pub struct InMemory {
+    /// The prepared schema.
+    ///
+    /// Note this is not the `projected_schema`, which is the schema
+    /// after applying column projections.
+    prepared_schema: SchemaRef,
+    /// The in-memory batches.
+    data: Arc<InMemoryBatches>,
+}
+
+impl InMemory {
+    pub fn new(queryable: bool, schema: SchemaRef) -> error_stack::Result<Self, Error> {
+        let data = Arc::new(InMemoryBatches::new(queryable, schema.clone()));
+        let source = Self {
+            prepared_schema: schema,
+            data,
+        };
+        Ok(source)
+    }
+
+    /// Add a batch, publishing it to the subscribers.
+    pub async fn add_batch(&self, batch: RecordBatch) -> error_stack::Result<(), Error> {
+        self.data.add_batch(batch).await
+    }
+}
+
+impl Source for InMemory {
+    fn schema(&self) -> SchemaRef {
+        self.prepared_schema.clone()
+    }
+
+    fn read(
+        &self,
+        read_config: ReadConfig,
+    ) -> futures::stream::BoxStream<'_, error_stack::Result<Batch, sparrow_interfaces::Error>> {
+        // TODO: Do we get any benefit from projecting the schema here?
+        // Or should the prepared schema just match the projected for in-memory sources?
+        let input_stream = if read_config.keep_open {
+            self.data
+                .subscribe()
+                .map_err(|e| {
+                    e.change_context(sparrow_interfaces::Error::internal_msg("invalid input"))
+                })
+                .and_then(move |batch| {
+                    let projected_schema = read_config.projected_schema.clone();
+                    async move {
+                        debug_assert_eq!(batch.schema(), projected_schema, "expected projected schema to equal batch schema. Saw schema: {:?}, expected projected: {:?}", batch.schema(), projected_schema);
+                        Batch::try_new_from_batch(batch).change_context(
+                            sparrow_interfaces::Error::internal_msg("invalid input"),
+                        )
+                    }
+                })
+                .boxed()
+        } else if let Some(batch) = self.data.current() {
+            futures::stream::once(async move {
+                Batch::try_new_from_batch(batch)
+                    .change_context(sparrow_interfaces::Error::internal_msg("invalid input"))
+            })
+            .boxed()
+        } else {
+            futures::stream::empty().boxed()
+        };
+
+        input_stream
+    }
+}
+
+/// Struct for managing in-memory batches.
+#[derive(Debug)]
+pub struct InMemoryBatches {
```

Comment:
No changes to this, just moved over.

Reply:
Does this need to be public? Ideally, the niteractions would happen through the `InMemory` struct that provides the synchronized access.

Reply:
It does not. Changed